### PR TITLE
Implement batch normalization at the code level

### PR DESCRIPTION
**Pull request by danpovey**
Includes some deeper modifications to the nnet3 framework-- the concept of 'memos' that can be passed between the Propagate and the corresponding Backprop. 

@tomkocse will work on testing this out with some script level changes, and seeing how it compares in terms of performance with using NormalizeLayer.

### CODE REVIEW

**Comment by danpovey** [comment]
Closing PR, as I'll be merging together with convolution.
